src/features/document/use_cases/add_document_use_case.py

```diff
@@ -1,28 +1,220 @@
+            blueprint_provider=document_service.get_blueprint,
+            node_attribute=BlueprintAttribute(name=target.attribute.name, attribute_type=entity.type),
+            recipe_provider=document_service.get_storage_recipes,
+        )
+        document_service.save(new_node, address.data_source, update_uncontained=update_uncontained)
+
+        return {"uid": f"{new_node.node_id}.{last_attribute_in_address}"}
+
+    if target.type != SIMOS.PACKAGE.value and target.type != "object":
+        validate_entity(
+            document,
+            document_service.get_blueprint,
+            document_service.get_blueprint(target.attribute.attribute_type),
+            "extend",
+        )
+
+    new_node = tree_node_from_dict(
+        {**document},
+        blueprint_provider=document_service.get_blueprint,
+        node_attribute=BlueprintAttribute(name=target.attribute.name, attribute_type=entity.type),
+        recipe_provider=document_service.get_storage_recipes,
+    )
+
+    if not target.is_array() and target.type != SIMOS.PACKAGE.value:
+        required_attribute_names = [attribute.name for attribute in new_node.blueprint.get_required_attributes()]
+        # If entity has a name, check if a file/attribute with the same name already exists on the target
+        if "name" in required_attribute_names and target.parent.duplicate_attribute(
+            new_node.entity.get("name", new_node.attribute.name)
+        ):
+            raise BadRequestException(
+                f"The document at '{address}' already has a child with name '{new_node.entity.get('name', new_node.attribute.name)}'"
+            )
+
+    if files:
+        merge_entity_and_files(new_node, files)
+
+    if target.type == SIMOS.PACKAGE.value:
+        target = target.children[0]  # Set target to be the packages content
+
+    if isinstance(target, ListNode) or target.parent.type == SIMOS.PACKAGE.value:
+        new_node.set_uid()
+        new_node.parent = target
+        new_node.key = str(len(target.children))
+        target.add_child(new_node)
+        document_service.save(target.find_parent(), address.data_source, update_uncontained=False)
+    else:
+        new_node.parent = target.parent
+        target.parent.replace(new_node.node_id, new_node)
+        document_service.save(target.find_parent(), address.data_source, update_uncontained=False)
+
+    document_service.save(new_node, address.data_source, update_uncontained=update_uncontained)
+
+    return {"uid": new_node.node_id}
 
 
 def add_document_use_case(
-    user: User,
     document: dict,
-    address: str,
+    address: Address,
+    document_service: DocumentService,
     files: Optional[List[UploadFile]] = None,
     update_uncontained: Optional[bool] = False,
-    repository_provider=get_data_source,
-):
-    document_service = DocumentService(repository_provider=repository_provider, user=user)
+) -> dict:
+    """Add document to a data source or existing entity.
```

Comment:
the use case is also used to add items to list
```suggestion
    """Add document to a data source or existing entity. Can also be used to add (complex) items to a list. 
```